src/software/simulation/physics/physics_world.cpp

```diff
@@ -1,78 +1,186 @@
+{
+    return physics_ball ? std::make_optional(physics_ball->getBallState()) : std::nullopt;
 }
 
-void PhysicsWorld::initWorld(const World& world)
+const std::vector<RobotStateWithId> PhysicsWorld::getRobotStates(
+    const TeamColour colour) const
 {
-    physics_field = std::make_shared<PhysicsField>(b2_world, world.field());
-    physics_ball =
-        std::make_shared<PhysicsBall>(b2_world, world.ball().currentState().ballState(),
-                                      BALL_MASS_KG, acceleration_due_to_gravity);
-    friendly_physics_robots.clear();
-    for (const auto& friendly_robot : world.friendlyTeam().getAllRobots())
+    std::vector<std::shared_ptr<PhysicsRobot>> physics_robots;
+    switch (colour)
     {
-        friendly_physics_robots.emplace_back(std::make_shared<PhysicsRobot>(
-            friendly_robot.id(), b2_world, friendly_robot.currentState().robotState(),
-            ROBOT_WITH_BATTERY_MASS_KG));
+        case TeamColour::BLUE:
+            physics_robots = blue_physics_robots;
+            break;
+        case TeamColour ::YELLOW:
+            physics_robots = yellow_physics_robots;
+            break;
     }
-    enemy_physics_robots.clear();
-    for (const auto& enemy_robot : world.enemyTeam().getAllRobots())
+
+    std::vector<RobotStateWithId> robot_states;
+    for (const auto& robot : physics_robots)
     {
-        enemy_physics_robots.emplace_back(std::make_shared<PhysicsRobot>(
-            enemy_robot.id(), b2_world, enemy_robot.currentState().robotState(),
-            ROBOT_WITH_BATTERY_MASS_KG));
+        if (!robot)
+        {
+            LOG(FATAL) << "Encountered a nullptr to a " << toString(colour)
+                       << " physics robot in the physics world";
+        }
+
+        auto state_with_id = RobotStateWithId{.id          = robot->getRobotId(),
+                                              .robot_state = robot->getRobotState()};
+        robot_states.emplace_back(state_with_id);
     }
-    current_timestamp = world.getMostRecentTimestamp();
+
+    return robot_states;
+}
+
+const std::vector<RobotStateWithId> PhysicsWorld::getYellowRobotStates() const
+{
+    return getRobotStates(TeamColour::YELLOW);
+}
+
+const std::vector<RobotStateWithId> PhysicsWorld::getBlueRobotStates() const
+{
+    return getRobotStates(TeamColour::BLUE);
+}
+
+const Timestamp PhysicsWorld::getTimestamp() const
+{
+    return current_timestamp;
+}
+
+void PhysicsWorld::setBallState(const BallState& ball_state)
+{
+    physics_ball = std::make_shared<PhysicsBall>(b2_world, ball_state, BALL_MASS_KG,
+                                                 acceleration_due_to_gravity);
 }
 
-World PhysicsWorld::getWorld() const
+void PhysicsWorld::removeBall()
 {
-    World new_world;
-    new_world.updateTimestamp(current_timestamp);
-    if (physics_ball)
+    physics_ball.reset();
+}
+
+void PhysicsWorld::addYellowRobots(const std::vector<RobotStateWithId>& robots)
```

Comment:
Would it be possible to move addXXXXRobots into a single function with an enum/boolean parameter for the team then wrap that in a addYellow/Blue robots function?

Just thinking it would probably minimize similar code between the functions and have the same interface.

Reply:
Ah so this is a bit trickier because I can't use the "regular" `switch` statement to select the correct `physics_robot` vector to add to. This is because I need a reference to the variable, not a copy. However you can't default-construct a reference variable, so code like this wouldn't work.

```
    std::vector<std::shared_ptr<PhysicsRobot>>& physics_robots;
    switch (colour)
    {
        case TeamColour::BLUE:
            physics_robots = blue_physics_robots;
            break;
        case TeamColour ::YELLOW:
            physics_robots = yellow_physics_robots;
            break;
    }
```

I could assign the variable directly with a ternary operator but then we lose the compile-time guarantee that all enum cases are handled. I think these functions are small enough there is little enough duplication we can tolerate it.